docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,80 @@
+---
+type: proposal
+title: Tenancy awareness in query path
+status: accepted
+owner: douglascamata
+menu: proposals-accepted
+---
+
+## Tenancy awareness in query path
+
+* **Owners:**
+  * [@douglascamata](https://github.com/douglascamata)
+
+* **Related Tickets:**
+  * [Querier: Added native multi-tenancy support](https://github.com/thanos-io/thanos/pull/4141).
+  * [Proposal: Native Multi-tenancy Support](https://github.com/thanos-io/thanos/pull/4055)
+
+* **Other docs:**
+  * [Multi-tenancy in Thanos](https://thanos.io/tip/operating/multi-tenancy.md/)
+
+This design doc proposes to add tenancy awareness in the query path.
+
+## Why
+
+In a multi-tenant environment, it's important to be able to identify which tenants are experiencing issues and configure (e.g. with different limits) each one of them individually and according to their usage of the platform so that the quality of service can be guaranteed to all the tenants.
+
+### Pitfalls of the current solution
+
+The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools or proxies. For example, it's impossible to determine which tenants are experiencing high latency or high error rates.
+
+## Goals
+
+* Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path. These features include, but aren't limited to, the following:
+  * Per-tenant observability.
+  * Per-tenant settings. For example, having different limits per tenant, which is a common request.
+  * Enforce presence of one or more tenant labels in queries.
```

Comment:
We can achieve these goals by having separate query instances for individual tenants. A valid goal would be simplifying management for multi-tenant setups and pooling resources to reduce waste.

Reply:
Hey @friedrichg, can you give us some good pointers for this implementation on the Cortex side? I'm not familiar with Cortex and its features, so I'm very unaware of how it implements things.

If it aligns well with how we want to move forward with this, it'll be great to reuse some code (be it importing or copying) so that both projects benefit from it. I know there are some design differences between the two projects, so let's see what can be done and reused.

Reply:
Yea, I think it's good point to mention Cortex way of dealing this as inspiration, but there is nothing wrong in revisiting this feature and checking if there is anything we could improve before implementing.

Ideally, we could have the same solution, so the tools that works on top of Cortex will work on top of Thanos. That's fair advantage to adopting similar patterns. Another advantage is easier migrations between projects, helpful if we will try to unify projects at some point.

Can we mention this somewhere in the doc (advantage and Cortex pattern)? Thanks!